File: src/detail-view.test.ts
import { describe, expect, test } from "bun:test";
import type { K8sObject } from "./kubectl.ts";
import {
    type Fact,
    type FactGroup,
    nodeAffinityLines,
    podLine,
    podView,
    referencesGroup,
    tolerationLines,
    topologySpreadLines,
    workloadView,
} from "./detail-view.ts";
import { NO_METRICS, metricsFromContainers } from "./usage.ts";

const iso = (minutesAgo: number) => new Date(Date.now() - minutesAgo * 60_000).toISOString();

const fact = (group: FactGroup, label: string): Fact | undefined => group.facts.find((f) => f.label === label);
const findFact = (groups: FactGroup[], label: string): Fact | undefined =>
    groups.flatMap((g) => g.facts).find((f) => f.label === label);

const POD: K8sObject = {
    kind: "Pod",
    metadata: {
        name: "web-abc",
        namespace: "demo",
        creationTimestamp: iso(60),
        labels: { app: "web" },
        annotations: { "kubectl.kubernetes.io/restartedAt": "2026-06-20T15:45:29+05:30" },
        ownerReferences: [{ kind: "ReplicaSet", name: "web-5f7", controller: true }],
    },
    spec: {
        nodeName: "node-1",
        serviceAccountName: "web-sa",
        restartPolicy: "Always",
        tolerations: [{ key: "node.kubernetes.io/not-ready", operator: "Exists", effect: "NoExecute", tolerationSeconds: 300 }],
        containers: [
            {
                name: "app",
                image: "example/app:1",
                ports: [{ name: "http", containerPort: 8080, protocol: "TCP" }],
                resources: { requests: { cpu: "25m", memory: "32Mi" }, limits: { cpu: "100m", memory: "96Mi" } },
            },
        ],
    },
    status: {
        phase: "Running",
        podIP: "10.0.0.9",
        hostIP: "192.168.1.5",
        qosClass: "Burstable",
        startTime: iso(60),
        conditions: [
            { type: "Initialized", status: "True" },
            { type: "Ready", status: "False", reason: "ContainersNotReady" },
        ],
        containerStatuses: [
            {
                name: "app",
                ready: true,
                started: true,
                restartCount: 2,
                state: { running: { startedAt: iso(30) } },
                lastState: { terminated: { reason: "OOMKilled", exitCode: 137, finishedAt: iso(31) } },
            },
        ],
    },
};

describe("podView", () => {
    const view = podView(POD, metricsFromContainers(new Map([["demo/web-abc/app", { cpu: "7m", memory: "40Mi" }]])));

    test("identity card carries owner, labels and annotations", () => {
        expect(fact(view.header, "Owner")?.ref).toEqual({ kind: "replicasets", name: "web-5f7", ns: "demo" });
        expect(fact(view.header, "Labels")?.chips).toEqual([{ k: "app", v: "web" }]);
        expect(fact(view.header, "Annotations")?.chips?.[0]?.k).toBe("kubectl.kubernetes.io/restartedAt");
    });

    test("node and service account are links, not text", () => {
        expect(findFact(view.groups, "Node")?.ref).toEqual({ kind: "nodes", name: "node-1" });
        expect(findFact(view.groups, "Service Account")?.ref).toEqual({
            kind: "serviceaccounts",
            name: "web-sa",
            ns: "demo",
        });
    });

    test("conditions are toned by status", () => {
        const chips = findFact(view.groups, "Conditions")?.chips ?? [];
        expect(chips[0]).toEqual({ v: "Initialized", tone: "ok" });
        expect(chips[1]).toEqual({ v: "Ready: ContainersNotReady", tone: "bad" });
    });

    test("the container card keeps the reason the last crash happened", () => {
        const c = view.containers[0]!;
        expect(c.state).toBe("Running");
        expect(c.stateTone).toBe("ok");
        expect(c.restarts).toBe(2);
        expect(c.restartReason).toBe("OOMKilled");
        expect(c.ports).toBe("http:8080/TCP");
        expect(c.cpu.used).toBeCloseTo(0.007);
        expect(c.cpu.limits).toBeCloseTo(0.1);
        expect(c.mem.used).toBe(40 * 1024 * 1024);
    });

    test("usage is unknown, not zero, without metrics", () => {
        const bare = podView(POD, NO_METRICS);
        expect(bare.containers[0]!.cpu.used).toBeNull();
        expect(bare.containers[0]!.cpu.requests).toBeCloseTo(0.025);
    });
});

describe("scheduling summaries", () => {
    test("tolerations read as key: effect", () => {
        expect(tolerationLines(POD.spec!)).toEqual(["node.kubernetes.io/not-ready: NoExecute for 300s"]);
    });

    test("node affinity flattens required and preferred terms", () => {
        const spec = {
            affinity: {
                nodeAffinity: {
                    requiredDuringSchedulingIgnoredDuringExecution: {
                        nodeSelectorTerms: [{ matchExpressions: [{ key: "disk", operator: "In", values: ["ssd"] }] }],
                    },
                    preferredDuringSchedulingIgnoredDuringExecution: [
                        { weight: 10, preference: { matchExpressions: [{ key: "zone", operator: "In", values: ["a"] }] } },
                    ],
                },
            },
        };
        expect(nodeAffinityLines(spec)).toEqual(["required: disk In ssd", "preferred (10): zone In a"]);
    });

    test("topology spread constraints keep skew and policy", () => {
        expect(
            topologySpreadLines({
                topologySpreadConstraints: [
                    { topologyKey: "kubernetes.io/hostname", maxSkew: 2, whenUnsatisfiable: "ScheduleAnyway" },
                ],
            }),
        ).toEqual(["kubernetes.io/hostname · maxSkew 2 · ScheduleAnyway"]);
    });
});

describe("podLine", () => {
    test("summarises a pod for the workload table", () => {
        const line = podLine(POD, NO_METRICS);
        expect(line.ready).toBe("1/1");
        expect(line.status).toBe("Running");
        expect(line.restarts).toBe(2);
        expect(line.lastRestartReason).toBe("OOMKilled");
        expect(line.node).toBe("node-1");
    });
});

describe("workloadView", () => {
    const deployment: K8sObject = {
        kind: "Deployment",
        metadata: { name: "web", namespace: "demo", creationTimestamp: iso(120), labels: { app: "web" } },
        spec: {
            replicas: 2,
            selector: { matchLabels: { app: "web" } },
            strategy: { type: "RollingUpdate", rollingUpdate: { maxSurge: "25%", maxUnavailable: "25%" } },
            template: {
                spec: {
                    containers: [
                        {
                            name: "app",
                            image: "example/app:1",
                            resources: { requests: { cpu: "25m" }, limits: { cpu: "100m" } },
                        },
                    ],
                },
            },
        },
        status: { replicas: 2, readyReplicas: 2, updatedReplicas: 2, availableReplicas: 1, conditions: [{ type: "Available", status: "True" }] },
    };

    const pods = [POD, { ...POD, metadata: { ...POD.metadata, name: "web-def" } } as K8sObject];
    const view = workloadView(
        deployment,
        "deployments",
        pods,
        metricsFromContainers(
            new Map([
                ["demo/web-abc/app", { cpu: "7m", memory: "40Mi" }],
                ["demo/web-def/app", { cpu: "3m", memory: "20Mi" }],
            ]),
        ),
    );

    test("rollout group carries the strategy knobs", () => {
        const rollout = view.groups[0]!;
        expect(rollout.title).toBe("Rollout");
        expect(fact(rollout, "Strategy")?.text).toBe("RollingUpdate");
        expect(fact(rollout, "Max Surge")?.text).toBe("25%");
        expect(fact(rollout, "Max Unavailable")?.text).toBe("25%");
        expect(fact(rollout, "Conditions")?.chips).toEqual([{ v: "Available", tone: "ok" }]);
    });

    test("counts come from the status, pods from the list", () => {
        expect(view.pods).toMatchObject({ desired: 2, updated: 2, ready: 2, available: 1 });
        expect(view.pods!.rows.map((r) => r.name)).toEqual(["web-abc", "web-def"]);
    });

    test("container usage is summed over the pods and says so", () => {
        expect(view.containers[0]!.cpu.used).toBeCloseTo(0.01);
        expect(view.containers[0]!.cpu.requests).toBeCloseTo(0.05); // 2 × 25m
        expect(view.containersNote).toBe("summed over 2 pods");
    });

    test("a single replica needs no aggregation note", () => {
        const single = workloadView(deployment, "deployments", [POD], NO_METRICS);
        expect(single.containersNote).toBeUndefined();
        expect(single.containers[0]!.cpu.requests).toBeCloseTo(0.025);
    });
});

describe("references", () => {
    const SPEC = {
        serviceAccountName: "web-sa",
        priorityClassName: "high",
        imagePullSecrets: [{ name: "registry-cred" }],
        volumes: [
            { name: "config", configMap: { name: "web-config" } },
            { name: "tls", secret: { secretName: "web-tls" } },
            { name: "data", persistentVolumeClaim: { claimName: "web-data" } },
            {
                name: "bundle",
                projected: { sources: [{ configMap: { name: "ca-bundle" } }, { secret: { name: "web-tls" } }] },
            },
            { name: "vault", csi: { nodePublishSecretRef: { name: "csi-cred" } } },
            { name: "share", cephfs: { secretRef: { name: "ceph-key" } } },
            { name: "scratch", emptyDir: {} },
        ],
        containers: [
            {
                name: "app",
                envFrom: [{ configMapRef: { name: "web-config" } }, { secretRef: { name: "app-env" } }],
                env: [
                    { name: "PLAIN", value: "x" },
                    { name: "DB_PASS", valueFrom: { secretKeyRef: { name: "db", key: "password" } } },
                    { name: "DB_HOST", valueFrom: { configMapKeyRef: { name: "db-conf", key: "host" } } },
                ],
            },
        ],
        initContainers: [{ name: "migrate", envFrom: [{ secretRef: { name: "db" } }] }],
    };

    const group = referencesGroup(SPEC, "demo", true)!;
    const refs = (label: string) => fact(group, label)?.refs ?? [];
    const names = (label: string) => refs(label).map((r) => r.name).sort();

    test("collects configmaps from volumes, projections, envFrom and env", () => {
        expect(names("ConfigMaps")).toEqual(["ca-bundle", "db-conf", "web-config"]);
    });

    test("collects secrets from every carrier the API server resolves", () => {
        expect(names("Secrets")).toEqual(["app-env", "ceph-key", "csi-cred", "db", "registry-cred", "web-tls"]);
    });

    test("claims are their own kind, and volumes with no reference are skipped", () => {
        expect(refs("Volume Claims")).toEqual([
            { kind: "persistentvolumeclaims", name: "web-data", ns: "demo", via: "volume data" },
        ]);
    });

    test("refs are namespaced so a link lands in the right namespace", () => {
        expect(refs("ConfigMaps").every((r) => r.ns === "demo")).toBe(true);
    });

    test("a name reached twice appears once, carrying both routes", () => {
        const cm = refs("ConfigMaps").find((r) => r.name === "web-config");
        expect(cm?.via).toBe("volume config, envFrom in app");
        const tls = refs("Secrets").find((r) => r.name === "web-tls");
        expect(tls?.via).toBe("volume tls, projected bundle");
    });

    test("env references name the variable, init containers name themselves", () => {
        expect(refs("Secrets").find((r) => r.name === "db")?.via).toBe("envFrom in migrate, env DB_PASS");
        expect(refs("ConfigMaps").find((r) => r.name === "db-conf")?.via).toBe("env DB_HOST");
    });

    test("many routes to one name are summarized rather than listed", () => {
        const many = referencesGroup(
            {
                containers: [
                    {
                        name: "app",
                        env: ["A", "B", "C", "D", "E"].map((n) => ({
                            name: n,
                            valueFrom: { secretKeyRef: { name: "big", key: n } },
                        })),
                    },
                ],
            },
            "demo",
        )!;
        expect(fact(many, "Secrets")?.refs?.[0]?.via).toBe("env A, env B +3 more");
    });

    test("identity adds links the workload has nowhere else to show", () => {
        expect(fact(group, "Service Account")?.ref).toEqual({ kind: "serviceaccounts", name: "web-sa", ns: "demo" });
        // Cluster-scoped: a namespace on this link would 404.
        expect(fact(group, "Priority Class")?.ref).toEqual({ kind: "priorityclasses", name: "high" });
        expect(fact(referencesGroup(SPEC, "demo")!, "Service Account")).toBeUndefined();
    });

    test("a spec that names nothing gets no group at all", () => {
        expect(referencesGroup({ containers: [{ name: "app", image: "x" }] })).toBeNull();
        expect(referencesGroup({})).toBeNull();
    });

    const deploy = (podSpec: Record<string, unknown>): K8sObject => ({
        kind: "Deployment",
        metadata: { name: "web", namespace: "demo", creationTimestamp: iso(10) },
        spec: { replicas: 1, template: { spec: podSpec } },
        status: {},
    });

    test("the workload page shows references above scheduling, and skips the group when empty", () => {
        const bare = workloadView(deploy({ containers: [{ name: "app", image: "x" }] }), "deployments", [], NO_METRICS);
        expect(bare.groups.map((g) => g.title)).toEqual(["Rollout", "Scheduling"]);

        const rich = workloadView(deploy(SPEC), "deployments", [], NO_METRICS);
        expect(rich.groups.map((g) => g.title)).toEqual(["Rollout", "References", "Scheduling"]);
    });

    test("a pod reads its own template's references and links its priority class", () => {
        const pod = { kind: "Pod", metadata: { name: "web-1", namespace: "demo" }, spec: SPEC } as K8sObject;
        const view = podView(pod, NO_METRICS);
        const group = view.groups.find((g) => g.title === "References")!;
        expect(group.facts.map((f) => f.label)).toEqual(["ConfigMaps", "Secrets", "Volume Claims"]);
        expect(findFact(view.groups, "Priority Class")?.ref).toEqual({ kind: "priorityclasses", name: "high" });
    });
});

File: src/internal/model/detailview_test.go
package model

import (
	"testing"
)

// src/detail-view.test.ts has no direct counterpart file in the Bun build (its
// coverage lived in src/detail-view.test.ts against the same helpers); these
// cover the behaviour the comments in detail-view.ts call out as load-bearing.

func factByLabel(g FactGroup, label string) *Fact {
	for i := range g.Facts {
		if g.Facts[i].Label == label {
			return &g.Facts[i]
		}
	}
	return nil
}

func groupByTitle(v DetailView, title string) *FactGroup {
	for i := range v.Groups {
		if v.Groups[i].Title == title {
			return &v.Groups[i]
		}
	}
	return nil
}

// Every site the API server resolves must be walked. A group that looks
// complete and is not teaches you to trust it.
func TestSpecRefsWalksEverySite(t *testing.T) {
	spec := map[string]any{
		"volumes": []any{
			map[string]any{"name": "cfg", "configMap": map[string]any{"name": "app-config"}},
			map[string]any{"name": "sec", "secret": map[string]any{"secretName": "app-secret"}},
			map[string]any{"name": "data", "persistentVolumeClaim": map[string]any{"claimName": "app-data"}},
			map[string]any{"name": "proj", "projected": map[string]any{"sources": []any{
				map[string]any{"configMap": map[string]any{"name": "proj-config"}},
				map[string]any{"secret": map[string]any{"name": "proj-secret"}},
			}}},
			map[string]any{"name": "csivol", "csi": map[string]any{
				"nodePublishSecretRef": map[string]any{"name": "csi-secret"}}},
			map[string]any{"name": "rbdvol", "rbd": map[string]any{
				"secretRef": map[string]any{"name": "rbd-secret"}}},
		},
		"imagePullSecrets": []any{map[string]any{"name": "regcred"}},
		"containers": []any{map[string]any{
			"name": "api",
			"envFrom": []any{
				map[string]any{"configMapRef": map[string]any{"name": "envfrom-config"}},
				map[string]any{"secretRef": map[string]any{"name": "envfrom-secret"}},
			},
			"env": []any{map[string]any{
				"name":      "DB_PASS",
				"valueFrom": map[string]any{"secretKeyRef": map[string]any{"name": "db-secret"}},
			}},
		}},
	}

	refs := SpecRefs(spec)
	got := map[string]string{}
	for _, r := range refs.Of(kindConfigMaps, "default") {
		got["cm/"+r.Name] = r.Via
	}
	for _, r := range refs.Of(kindSecrets, "default") {
		got["sec/"+r.Name] = r.Via
	}
	for _, r := range refs.Of(kindPVCs, "default") {
		got["pvc/"+r.Name] = r.Via
	}

	want := []string{
		"cm/app-config", "cm/proj-config", "cm/envfrom-config",
		"sec/app-secret", "sec/proj-secret", "sec/csi-secret", "sec/rbd-secret",
		"sec/regcred", "sec/envfrom-secret", "sec/db-secret",
		"pvc/app-data",
	}
	for _, key := range want {
		if _, ok := got[key]; !ok {
			t.Errorf("missing reference %s (found %v)", key, got)
		}
	}
	// The via carries WHERE it was reached from, which is the question the
	// group exists to answer.
	if got["cm/app-config"] != "volume cfg" {
		t.Errorf("via = %q, want %q", got["cm/app-config"], "volume cfg")
	}
	if got["sec/db-secret"] != "env DB_PASS" {
		t.Errorf("via = %q, want %q", got["sec/db-secret"], "env DB_PASS")
	}
}

// A name reached more than once appears once, carrying every route.
func TestSpecRefsDeduplicatesAndSummarizesVia(t *testing.T) {
	spec := map[string]any{
		"volumes": []any{
			map[string]any{"name": "a", "configMap": map[string]any{"name": "shared"}},
			map[string]any{"name": "b", "configMap": map[string]any{"name": "shared"}},
		},
		"containers": []any{map[string]any{
			"name":    "api",
			"envFrom": []any{map[string]any{"configMapRef": map[string]any{"name": "shared"}}},
		}},
	}
	list := SpecRefs(spec).Of(kindConfigMaps, "")
	if len(list) != 1 {
		t.Fatalf("expected 1 deduplicated ref, got %d", len(list))
	}
	if list[0].Via != "volume a, volume b, envFrom in api" {
		t.Errorf("via = %q", list[0].Via)
	}
}

func TestSpecRefsSummarizesBeyondThreeRoutes(t *testing.T) {
	vols := []any{}
	for _, n := range []string{"a", "b", "c", "d"} {
		vols = append(vols, map[string]any{"name": n, "configMap": map[string]any{"name": "shared"}})
	}
	list := SpecRefs(map[string]any{"volumes": vols}).Of(kindConfigMaps, "")
	if list[0].Via != "volume a, volume b +2 more" {
		t.Errorf("via = %q", list[0].Via)
	}
}

// Three empty dashes under a heading is worse than no heading.
func TestReferencesGroupIsNilWhenNothingIsNamed(t *testing.T) {
	if g := ReferencesGroup(map[string]any{"containers": []any{}}, "default", false); g != nil {
		t.Errorf("expected nil, got %+v", g)
	}
	// identity=true still produces a group when a service account is set.
	g := ReferencesGroup(map[string]any{"serviceAccountName": "api"}, "default", true)
	if g == nil || factByLabel(*g, "Service Account") == nil {
		t.Error("expected a References group carrying the service account")
	}
}

func TestPodViewSurfacesPendingAsWarn(t *testing.T) {
	pod := obj(t, `{"metadata":{"name":"web","namespace":"default"},
		"spec":{"containers":[{"name":"c","image":"nginx"}],
			"tolerations":[{"key":"node.kubernetes.io/not-ready","effect":"NoExecute","tolerationSeconds":300}]},
		"status":{"phase":"Pending"}}`)
	v := PodView(pod, NoMetrics)

	phase := factByLabel(v.Groups[0], "Phase")
	if phase == nil || phase.Text != "Pending" || phase.Tone != ToneWarn {
		t.Errorf("phase fact = %+v", phase)
	}
	sched := groupByTitle(v, "Scheduling")
	if sched == nil {
		t.Fatal("no Scheduling group")
	}
	tol := factByLabel(*sched, "Tolerations")
	if tol == nil || len(tol.Items) != 1 {
		t.Fatalf("tolerations = %+v", tol)
	}
	if tol.Items[0] != "node.kubernetes.io/not-ready: NoExecute for 300s" {
		t.Errorf("toleration line = %q", tol.Items[0])
	}
}

func TestPodContainerViewsCarryRestartReason(t *testing.T) {
	pod := obj(t, `{"metadata":{"name":"web","namespace":"default"},
		"spec":{"containers":[{"name":"api","image":"api:1",
			"resources":{"requests":{"cpu":"100m","memory":"64Mi"},
			             "limits":{"cpu":"500m","memory":"128Mi"}}}]},
		"status":{"containerStatuses":[{"name":"api","ready":true,"restartCount":3,
			"state":{"running":{"startedAt":"2026-01-01T00:00:00Z"}},
			"lastState":{"terminated":{"reason":"OOMKilled","exitCode":137,
				"finishedAt":"2026-01-01T00:00:00Z"}}}]}}`)

	views := PodContainerViews(pod, NoMetrics)
	if len(views) != 1 {
		t.Fatalf("expected 1 container, got %d", len(views))
	}
	c := views[0]
	if c.State != "Running" || c.StateTone != ToneOK {
		t.Errorf("state = %q/%q", c.State, c.StateTone)
	}
	if c.Restarts != 3 || c.RestartReason != "OOMKilled" {
		t.Errorf("restarts = %d, reason = %q", c.Restarts, c.RestartReason)
	}
	if c.CPU.Requests != 0.1 || c.CPU.Limits != 0.5 {
		t.Errorf("cpu gauge = %+v", c.CPU)
	}
	// No metrics-server means unknown, never a confident zero.
	if c.CPU.Used != nil {
		t.Errorf("cpu used = %v, want unknown", *c.CPU.Used)
	}
}

// Init containers are a MAX, not a sum: they never run alongside app containers.
func TestPodAllocationTreatsInitContainersAsMax(t *testing.T) {
	pod := obj(t, `{"spec":{
		"initContainers":[{"name":"init","resources":{"requests":{"cpu":"2","memory":"1Gi"}}}],
		"containers":[
			{"name":"a","resources":{"requests":{"cpu":"100m","memory":"64Mi"}}},
			{"name":"b","resources":{"requests":{"cpu":"200m","memory":"64Mi"}}}]}}`)
	a := PodAllocation(pod)
	// app containers sum to 300m, the init container wants 2 cores → 2 wins.
	if a.CPU.Requests != 2 {
		t.Errorf("cpu requests = %v, want 2", a.CPU.Requests)
	}
	if a.Mem.Requests != 1024*1024*1024 {
		t.Errorf("mem requests = %v, want 1Gi", a.Mem.Requests)
	}
}

func TestWorkloadViewSumsUsageAcrossPods(t *testing.T) {
	dep := obj(t, `{"kind":"Deployment","metadata":{"name":"api","namespace":"default"},
		"spec":{"replicas":2,"selector":{"matchLabels":{"app":"api"}},
			"strategy":{"type":"RollingUpdate","rollingUpdate":{"maxSurge":1,"maxUnavailable":0}},
			"template":{"spec":{"containers":[{"name":"api","image":"api:1",
				"resources":{"requests":{"cpu":"100m"}}}]}}},
		"status":{"replicas":2,"readyReplicas":2,"updatedReplicas":2,"availableReplicas":2}}`)
	pods := []Obj{
		*obj(t, `{"metadata":{"name":"api-1","namespace":"default"},
			"spec":{"containers":[{"name":"api"}]},"status":{"phase":"Running"}}`),
		*obj(t, `{"metadata":{"name":"api-2","namespace":"default"},
			"spec":{"containers":[{"name":"api"}]},"status":{"phase":"Running"}}`),
	}
	metrics := MetricsView(nil, map[string]Usage{
		"default/api-1/api": {CPU: "50m", Memory: "10Mi"},
		"default/api-2/api": {CPU: "70m", Memory: "20Mi"},
	})

	v := WorkloadView(dep, "deployments", pods, metrics)

	if v.ContainersNote != "summed over 2 pods" {
		t.Errorf("note = %q", v.ContainersNote)
	}
	if len(v.Containers) != 1 {
		t.Fatalf("containers = %d", len(v.Containers))
	}
	used := v.Containers[0].CPU.Used
	if used == nil {
		t.Fatal("expected summed usage")
	}
	// 50m + 70m = 120m = 0.12 cores
	if *used < 0.1199 || *used > 0.1201 {
		t.Errorf("summed cpu = %v, want ~0.12", *used)
	}
	// requests are per-pod × 2
	if v.Containers[0].CPU.Requests < 0.1999 || v.Containers[0].CPU.Requests > 0.2001 {
		t.Errorf("summed requests = %v, want ~0.2", v.Containers[0].CPU.Requests)
	}
	if v.Pods == nil || v.Pods.Desired != 2 || v.Pods.Ready != 2 {
		t.Errorf("pods block = %+v", v.Pods)
	}
	// Selector sits third in the header, after Age and Namespace.
	if v.Header.Facts[2].Label != "Selector" {
		t.Errorf("header order = %q at index 2", v.Header.Facts[2].Label)
	}
	rollout := groupByTitle(v, "Rollout")
	if rollout == nil || factByLabel(*rollout, "Max Surge") == nil {
		t.Error("expected a Rollout group with Max Surge")
	}
}

func TestNodeViewReportsCordonAndCapacity(t *testing.T) {
	node := obj(t, `{"metadata":{"name":"node-1"},
		"spec":{"unschedulable":true},
		"status":{"nodeInfo":{"kubeletVersion":"v1.35.1","osImage":"Ubuntu"},
			"capacity":{"cpu":"4","memory":"8Gi","pods":"110"},
			"allocatable":{"cpu":"3800m","memory":"7Gi","pods":"110"},
			"addresses":[{"type":"InternalIP","address":"192.168.1.10"}]}}`)
	pods := []Obj{*obj(t, `{"metadata":{"name":"p1","namespace":"default"},
		"spec":{"containers":[{"name":"c","resources":{"requests":{"cpu":"500m","memory":"512Mi"}}}]},
		"status":{"phase":"Running"}}`)}

	v := NodeView(node, pods, NoMetrics)

	sched := factByLabel(v.Groups[0], "Schedulable")
	if sched == nil || sched.Text != "Cordoned" || sched.Tone != ToneWarn {
		t.Errorf("schedulable = %+v", sched)
	}
	cap := groupByTitle(v, "Capacity")
	if cap == nil {
		t.Fatal("no Capacity group")
	}
	if got := factByLabel(*cap, "CPU").Text; got != "3800m allocatable of 4" {
		t.Errorf("cpu = %q", got)
	}
	if got := factByLabel(*cap, "Pods").Text; got != "1 of 110" {
		t.Errorf("pods = %q", got)
	}
	if got := factByLabel(*cap, "Requested").Text; got != "0.50 cores · 512Mi" {
		t.Errorf("requested = %q", got)
	}
}
